core/client/tracker/tracker.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tracker verifies consistency proofs
+package tracker
+
+import (
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/trillian/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+	tpb "github.com/google/trillian"
+)
+
+// UpdateTrustedPredicate return a bool indicating whether the local reference
+// for the latest SignedLogRoot should be updated.
+type UpdateTrustedPredicate func(cntRoot, newRoot types.LogRootV1) bool
+
+// LogRootVerifier verifies a Trillian Log Root.
+type LogRootVerifier interface {
+	VerifyRoot(trusted *types.LogRootV1, newRoot *tpb.SignedLogRoot, consistency [][]byte) (*types.LogRootV1, error)
+}
+
+// LogTracker keeps a continuous series of consistent log roots.
+type LogTracker struct {
+	trusted       types.LogRootV1
+	v             LogRootVerifier
+	updateTrusted UpdateTrustedPredicate
+	mu            sync.RWMutex
+}
+
+// New creates a log tracker from no trusted root.
+func New(lv LogRootVerifier) *LogTracker {
+	return NewFromSaved(lv, types.LogRootV1{})
+}
+
+// NewFromSaved creates a log tracker from a previously saved trusted root.
+func NewFromSaved(lv LogRootVerifier, lr types.LogRootV1) *LogTracker {
+	return &LogTracker{v: lv, trusted: lr, updateTrusted: isNewer}
+}
+
+// LastVerifiedLogRoot retrieves the tree size of the latest log root.
+func (l *LogTracker) LastVerifiedLogRoot() *pb.LogRootRequest {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+	return l.logRootRequest()
+}
+
+func (l *LogTracker) logRootRequest() *pb.LogRootRequest {
+	return &pb.LogRootRequest{
+		TreeSize: int64(l.trusted.TreeSize),
+		RootHash: l.trusted.RootHash,
+	}
+}
+
+// VerifyLogRoot verifies root and updates the trusted root if it is newer.
+// req must be equal to the most recent value from LastVerifiedLogRoot().
+// If two clients race to VerifyLogRoot at the same time, if one of them updates the root, the other will fail.
```

Comment:
If the verification fails, it is critical, but if it's just a race then it looks recoverable for the clients? Do you handle this race thing in clients?

Reply:
The verification is a critical failure.  
The previous client would block trying to obtain a mutex and timeout. 
This implementation will allow clients to make calls optimistically and then fail the verification if things get out of sync. 

In its current state, the tracker and client libraries are not designed to be used in a multi-threaded context.   However, this is changing with #1311.  

The Next PR will try to give nieve multi-threaded support by *not* updating the trusted root.